File: negative_keyword_analysis.R
# ============================================================
# Negative Keyword Analysis: Word-Level Post-Click Behavior
# R Implementation
# ============================================================
# Author: Elizabeth Lawson
# Description:
#   Identifies potential negative keywords by breaking paid search
#   queries into individual words and aggregating GA post-click
#   behavioral metrics (bounce rate, sessions, cost) at the word level.
#
#   This version pulls data directly from Google Analytics via the
#   Google Analytics Reporting API (Universal Analytics / GA3).
#   Because Google Ads is connected to GA, no session-level join
#   is required — cost and click data flow directly into GA sessions.
#
# Requirements:
#   install.packages(c("googleAnalyticsR", "stopwords", "qdap"))
#
# Setup:
#   1. Update DATE_RANGE to your desired analysis window
#   2. Update VIEW_NAME_PATTERN to match your GA view name
#   3. Update BRAND_TERMS, EXCLUDED_TERMS with your own lists
#      (or load them from Google Sheets — see notes below)
#   4. Update OUTPUT_PATH to your desired output location
#
# Notes:
#   - This script was originally written for Universal Analytics (GA3).
#     For GA4, use the Python version or SQL version in this repo.
#   - The R version includes more sophisticated preprocessing:
#     geographic term removal, brand detection, and Spanish stopwords.
#     These enhancements can be adapted to the Python/SQL versions.
# ============================================================


# ============================================================
# Configuration
# ============================================================

# Date range for analysis
DATE_RANGE <- c("2024-01-01", "2024-01-31")  # Update to your desired range

# Pattern to match your GA view name — update to match your account
VIEW_NAME_PATTERN <- "your view name"

# Output file path — update to your desired output location
OUTPUT_PATH <- "~/negative_keyword_output.csv"

# Bounce rate threshold for flagging words
BOUNCE_RATE_THRESHOLD <- 0.50

# Minimum sessions for a word to be included
MIN_SESSIONS <- 3

# ============================================================
# Brand/term lists
# These can be loaded from Google Sheets (see commented code below)
# or defined directly as vectors here.
# ============================================================

# Terms to exclude entirely from analysis
# Add industry-specific terms that are part of your targeting
# but should not be flagged as negative keywords
EXCLUDED_TERMS <- c(
  # "your_excluded_term_1",
  # "your_excluded_term_2"
)

# Brand terms — words that identify which brand/account the query belongs to
# Used to associate queries with the correct brand when it's not explicit
BRAND_TERMS <- data.frame(
  term  = c("your_brand_term"),   # the word as it appears in queries
  brand = c("your_brand_name")    # the brand it maps to
)

# Multi-word phrases to protect from splitting
# These will have spaces replaced with underscores before tokenization
PROTECTED_PHRASES <- list(
  # c("multi word phrase", "multi_word_phrase")
)


# ============================================================
# Uncomment to load reference lists from Google Sheets instead
# ============================================================
# library(gsheet)
#
# excluded_url <- 'https://docs.google.com/spreadsheets/d/YOUR_SHEET_ID/edit#gid=0'
# EXCLUDED_TERMS <- data.frame(gsheet2tbl(excluded_url))
#
# brands_url <- 'https://docs.google.com/spreadsheets/d/YOUR_SHEET_ID/edit#gid=1'
# BRAND_TERMS <- data.frame(gsheet2tbl(brands_url))


# ============================================================
# Load libraries
# ============================================================

library(googleAnalyticsR)
library(stopwords)
library(qdap)


# ============================================================
# Authenticate and get data from Google Analytics
# ============================================================

# Authenticate — opens browser for Google login
ga_auth(new_user = TRUE)

# Find your view ID
account_list <- ga_account_list()
where <- grep(VIEW_NAME_PATTERN, account_list$viewName, ignore.case = TRUE)
ViewID <- account_list[where, ]$viewId

if (length(ViewID) == 0) {
  stop("No view found matching: ", VIEW_NAME_PATTERN,
       "\nAvailable views:\n",
       paste(account_list$viewName, collapse = "\n"))
}

cat("Using view:", account_list$viewName[where], "\n")

# Define filters
# Only include sessions with bounces, clicks, and a matched query
mf  <- met_filter("bounces",   "GREATER_THAN", 0)
mf2 <- met_filter("sessions",  "GREATER",      0)
mf3 <- met_filter("adclicks",  "GREATER",      0)
fc  <- filter_clause_ga4(list(mf, mf2, mf3), operator = "AND")

df  <- dim_filter("admatchedQuery", "EXACT", "(not set)",     not = TRUE)
df2 <- dim_filter("admatchedQuery", "EXACT", "your_brand",    not = TRUE)  # exclude branded queries
fc2 <- filter_clause_ga4(list(df, df2), operator = "AND")

# Pull data from GA
ga_data <- google_analytics(
  viewId      = ViewID,
  date_range  = DATE_RANGE,
  dimensions  = c("adGroup", "admatchedQuery", "keyword"),
  metrics     = c("adClicks", "sessions", "bounces", "adCost"),
  met_filters = fc,
  dim_filters = fc2,
  max         = 200000,
  anti_sample = TRUE
)

cat("Loaded", nrow(ga_data), "queries\n")


# ============================================================
# Preprocessing: protect multi-word phrases
# ============================================================

for (phrase in PROTECTED_PHRASES) {
  ga_data$admatchedQuery <- gsub(
    phrase[1], phrase[2], ga_data$admatchedQuery
  )
}


# ============================================================
# Tokenize: break each query into individual words
# ============================================================

long_data <- NULL
itime <- proc.time()[3]

for (i in 1:nrow(ga_data)) {

  words <- rm_stopwords(ga_data$admatchedQuery[i], " ")[[1]]

  # Remove state names and abbreviations (geographic noise)
  words <- words[!words %in% tolower(state.abb)]
  words <- words[!words %in% tolower(state.name)]

  # Remove keyword terms (words already being targeted)
  words <- words[!words %in% gsub(
    "[+]", "", tolower(strsplit(ga_data$keyword[i], " ")[[1]])
  )]

  # Remove excluded terms
  if (length(EXCLUDED_TERMS) > 0) {
    words <- words[!words %in% tolower(EXCLUDED_TERMS)]
  }

  # Remove brand terms
  words <- words[!words %in% tolower(BRAND_TERMS$term)]

  # Remove English and Spanish stopwords
  words <- words[!words %in% tolower(stopwords(language = "en"))]
  words <- words[!words %in% tolower(stopwords(language = "es"))]

  # Remove numbers
  words <- words[!grepl("[[:digit:]]", words)]

  # Remove punctuation
  words <- gsub("[[:punct:]]", "", words)
  words <- words[words != ""]

  if (length(words) > 0) {
    temp <- data.frame(
      words    = words,
      AdGroup  = ga_data$adGroup[i],
      Query    = ga_data$admatchedQuery[i],
      keyword  = ga_data$keyword[i],
      adClicks = ga_data$adClicks[i],
      sessions = ga_data$sessions[i],
      bounces  = ga_data$bounces[i],
      adCost   = ga_data$adCost[i],
      stringsAsFactors = FALSE
    )
    long_data <- rbind(long_data, temp)
  }

  if (i %% 100 == 0) {
    ctime     <- proc.time()[3]
    timetoend <- ((ctime - itime) / i) * (nrow(ga_data) - i)
    cat(i, "of", nrow(ga_data),
        "| Est. time remaining:", round(timetoend / 60, 1), "min\n")
  }
}

# Remove any remaining rows with numbers in words
long_data <- long_data[!grepl("[[:digit:]]", long_data$words), ]
long_data <- long_data[long_data$words != "", ]

cat("Tokenized into", nrow(long_data), "word-level rows\n")


# ============================================================
# Aggregate to word level
# ============================================================

agg <- aggregate(
  long_data[, c("adClicks", "sessions", "bounces", "adCost")],
  by = list(
    words   = long_data$words,
    AdGroup = long_data$AdGroup,
    query   = long_data$Query,
    keyword = long_data$keyword
  ),
  sum
)

agg$BounceRate <- agg$bounces / agg$sessions
agg <- agg[agg$words != "", ]
agg <- agg[order(agg$BounceRate, decreasing = TRUE), ]

cat("Aggregated to", nrow(agg), "word-level rows\n")


# ============================================================
# Flag high-bounce words
# ============================================================

high_bounce <- agg[
  agg$BounceRate >= BOUNCE_RATE_THRESHOLD &
  agg$sessions   >= MIN_SESSIONS,
]

cat("Flagged", nrow(high_bounce), "words with bounce rate >=",
    BOUNCE_RATE_THRESHOLD, "and >=", MIN_SESSIONS, "sessions\n")

print(head(high_bounce[, c("words", "sessions", "BounceRate", "adCost")], 10))


# ============================================================
# Save output
# ============================================================

write.csv(agg, OUTPUT_PATH, row.names = FALSE)
cat("Output saved to:", OUTPUT_PATH, "\n")

File: Negative keyword analysis.py
# ============================================================
# Negative Keyword Analysis: Word-Level Post-Click Behavior
# Python Implementation (Manual / No BigQuery Required)
# ============================================================
# Author: Elizabeth Lawson
# Description:
#   Identifies potential negative keywords by breaking paid search
#   queries into individual words and aggregating GA4 post-click
#   behavioral metrics (bounce rate, sessions, cost) at the word level.
#
#   This version uses a manually exported GA4 CSV report.
#   No BigQuery or data engineering setup required.
#
# Requirements:
#   pip install pandas nltk
#
# Data Setup:
#   Export a custom GA4 Explore report with the following fields:
#     Dimensions: Session Google Ads query
#     Metrics:    Sessions, Engaged sessions, Google Ads cost
#   Save the export as a CSV file and update DATA_PATH below.
#
#   A sample dataset (sample_data.csv) is included in this repo
#   so you can run the analysis immediately without your own data.
#
# Usage:
#   1. Update DATA_PATH to point to your exported CSV
#   2. Adjust BOUNCE_RATE_THRESHOLD and MIN_SESSIONS as needed
#   3. Run the script
#   4. Review output in high_bounce_words.csv and flagged_queries.csv
# ============================================================

import pandas as pd
import nltk
from nltk.corpus import stopwords

nltk.download('stopwords')

# ============================================================
# Configuration — update these values for your use case
# ============================================================

# Path to your GA4 export CSV
# Use 'sample_data.csv' to run with the included sample dataset
DATA_PATH = 'sample_data.csv'

# Number of header rows to skip in the GA4 export
# GA4 exports typically include 6 rows of metadata before the data
SKIP_ROWS = 0  # Set to 6 if using a raw GA4 export

# Words that should be kept together as a single token
# Add multi-word terms specific to your industry
# Format: {'original phrase': 'replacement_with_underscores'}
PROTECTED_PHRASES = {
    # 'myasthenia gravis': 'myasthenia_gravis',
    # 'add your': 'add_your',
    # 'multi word terms': 'multi_word_terms',
}

# Minimum sessions for a word to be included in output
# Increase this to filter out low-volume noise
MIN_SESSIONS = 3

# Bounce rate threshold for flagging words as potential negatives
# Words above this threshold with sufficient sessions will be flagged
BOUNCE_RATE_THRESHOLD = 0.50

# Output file paths
OUTPUT_WORDS_PATH = 'high_bounce_words.csv'
OUTPUT_QUERIES_PATH = 'flagged_queries.csv'


# ============================================================
# Load and prepare data
# ============================================================

df = pd.read_csv(DATA_PATH, skiprows=SKIP_ROWS)

# Remove unnamed columns (common in GA4 exports)
df = df.loc[:, ~df.columns.str.contains('^Unnamed')]

# Remove rows with no query data
df = df.dropna(subset=['Session Google Ads query'])

# Apply protected phrase replacements before tokenizing
for phrase, replacement in PROTECTED_PHRASES.items():
    df['Session Google Ads query'] = df['Session Google Ads query'].str.replace(
        phrase, replacement, case=False
    )

# Calculate bounces and bounce rate
# GA4 uses "Engaged sessions" — bounces = sessions that were not engaged
df['Bounces'] = df['Sessions'] - df['Engaged sessions']
df['Bounce Rate'] = df['Bounces'] / df['Sessions']

print(f"Loaded {len(df):,} queries")
print(df.head())


# ============================================================
# Tokenize: break each query into individual words
# ============================================================

stop_words = set(stopwords.words('english'))

long_data_rows = []

for _, row in df.iterrows():
    query = str(row['Session Google Ads query'])

    # Split query into words, removing stopwords
    words = [
        word for word in query.split()
        if word.lower() not in stop_words
        and word.strip() != ''
    ]

    for word in words:
        long_data_rows.append({
            'Session Google Ads query': query,
            'Word': word.lower(),
            'Google Ads cost': row.get('Google Ads cost', 0),
            'Sessions': row['Sessions'],
            'Bounces': row['Bounces'],
            'Bounce Rate': row['Bounce Rate']
        })

long_data = pd.DataFrame(long_data_rows)

# Ensure Sessions is numeric
long_data['Sessions'] = pd.to_numeric(
    long_data['Sessions'], errors='coerce'
).fillna(0).astype(int)

# Filter out zero-session rows
long_data = long_data[long_data['Sessions'] > 0]

print(f"\nTokenized into {len(long_data):,} word-level rows")


# ============================================================
# Aggregate to word level
# ============================================================

word_data = long_data.groupby('Word').agg(
    Google_Ads_Cost=('Google Ads cost', 'sum'),
    Sessions=('Sessions', 'sum'),
    Bounces=('Bounces', 'sum')
).reset_index()

word_data['Bounce Rate'] = word_data['Bounces'] / word_data['Sessions']
word_data = word_data[word_data['Sessions'] > 0]
word_data = word_data.sort_values(
    by=['Bounce Rate', 'Sessions'], ascending=[False, False]
)

print(f"\nAggregated to {len(word_data):,} unique words")


# ============================================================
# Flag high-bounce words
# ============================================================

high_bounce = word_data[
    (word_data['Bounce Rate'] >= BOUNCE_RATE_THRESHOLD) &
    (word_data['Sessions'] >= MIN_SESSIONS)
].copy()

print(f"\nFlagged {len(high_bounce):,} words with bounce rate >= "
      f"{BOUNCE_RATE_THRESHOLD:.0%} and >= {MIN_SESSIONS} sessions")
print(high_bounce.head(10).to_string(index=False))

high_bounce.to_csv(OUTPUT_WORDS_PATH, index=False)
print(f"\nWord-level output saved to: {OUTPUT_WORDS_PATH}")


# ============================================================
# Drill into flagged words — show associated queries
# ============================================================

flagged_words = set(high_bounce['Word'].str.lower())

flagged_queries = long_data[
    long_data['Word'].str.lower().isin(flagged_words)
].copy()

flagged_queries = flagged_queries.sort_values(
    by=['Sessions', 'Bounce Rate'], ascending=[False, False]
)

flagged_queries.to_csv(OUTPUT_QUERIES_PATH, index=False)
print(f"Query-level detail saved to: {OUTPUT_QUERIES_PATH}")


# ============================================================
# Helper: investigate a specific word
# ============================================================

def investigate_word(word):
    """
    Print all queries containing a specific flagged word,
    along with their sessions, bounce rate, and cost.
    
    Usage: investigate_word('your_word_here')
    """
    subset = long_data[long_data['Word'].str.lower() == word.lower()]
    if subset.empty:
        print(f"No queries found containing '{word}'")
        return

    print(f"\n=== Queries containing '{word}' ===")
    print(f"Total sessions: {subset['Sessions'].sum():,}")
    print(f"Overall bounce rate: "
          f"{subset['Bounces'].sum() / subset['Sessions'].sum():.1%}")
    print(f"Total cost: ${subset['Google Ads cost'].sum():,.2f}")
    print(f"\nIndividual queries:")
    print(subset[['Session Google Ads query', 'Sessions',
                  'Bounce Rate', 'Google Ads cost']]
          .sort_values('Sessions', ascending=False)
          .to_string(index=False))

# Example usage:
# investigate_word('lamborghini')
# investigate_word('craigslist')
